source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.h

```diff
@@ -0,0 +1,219 @@
+  };
+
+  // Factory used to create worker-local load balancer on the worker thread.
+  class WorkerLocalLbFactory : public Upstream::LoadBalancerFactory {
+  public:
+    WorkerLocalLbFactory(OptRef<const Upstream::LoadBalancerConfig> lb_config,
+                         const Upstream::ClusterInfo& cluster_info,
+                         const Upstream::PrioritySet& priority_set, Runtime::Loader& runtime,
+                         Envoy::Random::RandomGenerator& random, TimeSource& time_source)
+        : lb_config_(lb_config), cluster_info_(cluster_info), priority_set_(priority_set),
+          runtime_(runtime), random_(random), time_source_(time_source) {}
+
+    Upstream::LoadBalancerPtr create(Upstream::LoadBalancerParams params) override;
+
+    bool recreateOnHostChange() const override { return false; }
+
+  protected:
+    OptRef<const Upstream::LoadBalancerConfig> lb_config_;
+
+    const Upstream::ClusterInfo& cluster_info_;
+    const Upstream::PrioritySet& priority_set_;
+    Runtime::Loader& runtime_;
+    Envoy::Random::RandomGenerator& random_;
+    TimeSource& time_source_;
+  };
+
+public:
+  ClientSideWeightedRoundRobinLoadBalancer(OptRef<const Upstream::LoadBalancerConfig> lb_config,
+                                           const Upstream::ClusterInfo& cluster_info,
+                                           const Upstream::PrioritySet& priority_set,
+                                           Runtime::Loader& runtime,
+                                           Envoy::Random::RandomGenerator& random,
+                                           TimeSource& time_source);
+
+private:
+  friend class ClientSideWeightedRoundRobinLoadBalancerFriend;
+
+  // {Upstream::ThreadAwareLoadBalancer} Interface implementation.
+  Upstream::LoadBalancerFactorySharedPtr factory() override { return factory_; }
+  absl::Status initialize() override;
+
+  // Initialize LB based on the config.
+  void initFromConfig(const ClientSideWeightedRoundRobinLbConfig& lb_config);
+
+  // Start weight updates on the main thread.
+  void startWeightUpdatesOnMainThread(Event::Dispatcher& main_thread_dispatcher);
+
+  // Update weights using client side host LB policy data for all priority sets.
+  // Executed on the main thread.
+  void updateWeightsOnMainThread();
+
+  // Update weights using client side host LB policy data for all `hosts`.
+  void updateWeightsOnHosts(const HostVector& hosts);
+
+  // Add client side host LB policy data to all `hosts`.
+  static void addClientSideLbPolicyDataToHosts(const HostVector& hosts);
+
+  // Get weight based on client side host LB policy data if it is valid (not
+  // empty at least since `min_non_empty_since` and updated no later than
+  // `max_last_update_time`), otherwise return std::nullopt.
+  static absl::optional<uint32_t>
+  getClientSideWeightIfValidFromHost(const Host& host, const MonotonicTime& min_non_empty_since,
+                                     const MonotonicTime& max_last_update_time);
```

Comment:
`MonotonicTime` rather than `const MonotonicTime&`. Timepoint basically is an integer, const reference make no sense.